lib/README.md

```diff
@@ -0,0 +1,144 @@
+# Gutenberg PHP
+
+## File structure
+
+Gutenberg adds features to WordPress Core using PHP hooks and filters. Some
+features, once considered stable and useful, are merged into Core during a Core
+release. Some features remain in the plugin forever or are removed.
+
+To make it easier for contributors to know which features need to be merged to
+Core and which features can be deleted, Gutenberg uses the following file
+structure for its PHP code:
+
+- `lib/experimental` - Experimental features that exist only in the plugin. They
+  are not ready to be merged to Core.
+- `lib/stable` - Stable features that exist only in the plugin. They could one
+  day be merged to Core, but not yet.
+- `lib/compat/wordpress-X.Y` - Stable features that are intended to be merged to
+  Core in the future X.Y release, or that were previously merged to Core in the
+  X.Y release and remain in the plugin for backwards compatibility when running
+  the plugin on older versions of WordPress.
+
+## Best practices
+
+### Prefer the `wp` prefix
+
+For features that may be merged to Core, it's best to use a `wp_` prefix for functions or a `WP_` prefix for classes. 
+
+This applies to both experimental and stable features.
+
+Using the `wp_` prefix avoids us having to rename functions and classes from `gutenberg_` to `wp_` if the feature is merged to Core.
+
+Functions that are intended solely for the plugin, e.g., plugin infrastructure, should use the `gutenberg_` prefix.
+
+#### Feature that might be merged to Core
+
+```php
+function wp_get_navigation( $slug ) { ... }
+```
+
+#### Plugin infrastructure that will never be merged to Core
+
+```php
+function gutenberg_get_navigation( $slug ) { ... }
+```
+
+### Group PHP code by _feature_
```

Comment:
I actually think the `lib/compat` folder should group its code by "core" php file. Meaning if the change needs to happen in `edit-form-blocks.php` in 5.9, I'd put it in `lib/compat/5.9/edit-form-blocks.php` and for new files, try to come up with the best file name that will suite core merge. In practice, this is also very close to "group by feature".

The main goal is to simplify the work for the person back porting.

Reply:
The downside of that though is that you can't move features in and out of "experimental" status e.g. using `git mv lib/experimental/cool-new-feature.php lib/compat/wordpress-6.0/cool-new-feature.php`. Not sure if we do this in practice, mind you 😀